### PR TITLE
Fix a computational problem of scaledSoftmax.

### PR DESCRIPTION
**Pull request by yuanzexi**
**Problem**: The original implementation results in wrong results of sum of softmax such that the results of BERT models (128 < seq_len < 384 and seq_len > 384) are very large or even 'nan', especially for FP16 mode.
**Solution**: This implementation fix the computational problem such that the results of BERT models (128 < seq_len < 384 and seq_len > 384) become correct.

Signed-off-by: yuanzexi <hiyuanzexi@outlook.com>

### CODE REVIEW

File: plugin/common/common.cuh [hunk 1]
```diff
@@ -322,66 +322,54 @@ __device__ inline void scaledSoftmaxSmall(
...
-    __shared__ float rZ;
-    __shared__ float fMax;
-
-    const int offset = (blockIdx.y * gridDim.x + blockIdx.x) * ld;
-
-    const float w(rsqrtHeadSize);
-    cub::Sum sum;
-    float threadData(-FLT_MAX);
-
-    if (lastValid >= blockDim.x)
```
Comment (ttyio) [inline plugin/common/common.cuh:342]: Could we keep this check here in case the kernel is used in scenario `TPB > ld`? thanks!

**Comment by ttyio** [comment]
Adding @rajeevsrao for visibility.

Thanks @yuanzexi , good catch!

So you fixed 2 issues in this PR, right?
- The fMax is wrong when `TPB < ld` and there are multiple blocks
- The `- fMax` is missing in the final softmax calculation

The overall looks good to me. Could you follow https://github.com/NVIDIA/TensorRT/blob/master/CONTRIBUTING.md to reformat the code? thanks!

File: plugin/common/common.cuh (hunk 1 above)
Comment (yuanzexi) [reply plugin/common/common.cuh:342]: Yeah, you can keep it. Actually, this check confused me a lot. From my point of view, with this check, the initial value of `threadData` will be 0 in the scenario `TPB <= lastValid` but -FLT_MAX in the scenario `TPB > lastValid`. However, `threadData` will definitely be updated by `threadData = max(static_cast<float>(input[idx]), threadData);` in the next loop, so does this check is necessary for threads in the block? Or could we directly initialize the `threadData` by 0 instead of `-FLT_MAX`?

**Comment by yuanzexi** [comment]
> Adding @rajeevsrao for visibility.
> 
> Thanks @yuanzexi , good catch!
> 
> So you fixed 2 issues in this PR, right?
> 
> * The fMax is wrong when `TPB < ld` and there are multiple blocks
> * The `- fMax` is missing in the final softmax calculation
> 
> The overall looks good to me. Could you follow https://github.com/NVIDIA/TensorRT/blob/master/CONTRIBUTING.md to reformat the code? thanks!

Sorry about the code format, I'll format the code later.

File: plugin/common/common.cuh (hunk 1 above)
Comment (ttyio) [reply plugin/common/common.cuh:342]: the `threadData = max(static_cast<float>(input[idx])` is only called on condition `i < lastValid`, if `TPB <= lastValid`, we will accumulate the `-FLT_MAX` in next block reduced sum.
If we initialize the threadData to 0, then the `threadData = max(static_cast<float>(input[idx])` is invalid if the input is negative.

Comment (yuanzexi) [reply plugin/common/common.cuh:342]: From my point of view, maybe we just try to get the fmax here for `BlockReduceMax`? I'll set `threadData` as 0 before `BlockReduceSum`.

File: plugin/common/common.cuh [hunk 2]
```diff
@@ -322,66 +322,54 @@ __device__ inline void scaledSoftmaxSmall(
...
+    threadData = max(static_cast<float>(input[idx]), threadData);
+  }
+
+  const float maxElem = BlockReduce(tmpStorage).Reduce(threadData, cub::Max());
+  if (threadIdx.x == 0) {
+    fMax = maxElem;
+  }
+  __syncthreads();
+
+  threadData = 0;
```
Comment (yuanzexi) [inline plugin/common/common.cuh:353]: I'll set `threadData` as 0 here for all threads.

File: plugin/common/common.cuh [hunk 3]
```diff
@@ -322,66 +322,54 @@ __device__ inline void scaledSoftmaxSmall(
...
+  const float w(rsqrtHeadSize);
+  cub::Sum sum;
+  float threadData(-FLT_MAX);
+
+  for (int i = threadIdx.x; i < lastValid; i += TPB) {
+    const int idx = offset + i;
+    threadData = max(static_cast<float>(input[idx]), threadData);
+  }
+
+  const float maxElem = BlockReduce(tmpStorage).Reduce(threadData, cub::Max());
```
Comment (ttyio) [inline plugin/common/common.cuh:347]: some threadData is -FLT_MAX when the maxElem is computed here when `TPB > ld`

File: plugin/common/common.cuh (hunk 1 above)
Comment (ttyio) [reply plugin/common/common.cuh:342]: I mean line 347

File: plugin/common/common.cuh (hunk 3 above)
Comment (yuanzexi) [reply plugin/common/common.cuh]: Oh!  I see~ Thanks for answering!

**Comment by yuanzexi** [comment]
I have formated the code and the check we discussed has been added. Looking forward to your code review.

File: plugin/common/common.cuh [hunk 4]
```diff
@@ -35,7 +35,6 @@ __device__ inline T rsqrt(const T& x);
 template <typename T>
 __device__ inline T exp(const T& x);
 
-
```
Comment (ttyio) [inline plugin/common/common.cuh:38]: could you also revert the changes of blank lines, this could help us to do the integration between public repo and internal repo, thanks!

Comment (yuanzexi) [reply plugin/common/common.cuh]: No problem!  I have reverted them. Looking forward to your review.

File: plugin/common/common.cuh [hunk 5]
```diff
@@ -326,7 +326,7 @@ template <typename T, unsigned TPB>
 __device__ inline void scaledSoftmax(
     const int ld, const int lastValid, const float rsqrtHeadSize, const T* input, T* output)
 {
-
+
```
Comment (ttyio) [inline plugin/common/common.cuh:329]: could you also remove the blanks in this line? thanks

File: plugin/common/common.cuh [hunk 6]
```diff
@@ -343,10 +343,11 @@ __device__ inline void scaledSoftmax(
     {
         threadData = 0;
     }
+
```
Comment (ttyio) [inline plugin/common/common.cuh:346]: blank line

Comment (yuanzexi) [reply plugin/common/common.cuh]: removed.

File: plugin/common/common.cuh (hunk 5 above)
Comment (yuanzexi) [reply plugin/common/common.cuh]: removed.

**Comment by ttyio** [comment]
LGTM, thanks @yuanzexi 

Assign to @rajeevsrao , thanks!

**Review by ttyio — APPROVED** [review]

**Comment by rajeevsrao** [comment]
Thanks for the fix @yuanzexi - will test internally and integrate.
